File: android/src/main/java/com/damly/voice/synthesizer/IflytekSynthesizerModule.java
package com.damly.voice.synthesizer;

import android.content.Context;
import android.media.AudioManager;
import android.os.Bundle;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.iflytek.cloud.SpeechConstant;
import com.iflytek.cloud.SpeechError;
import com.iflytek.cloud.SpeechSynthesizer;
import com.iflytek.cloud.SpeechUtility;
import com.iflytek.cloud.SynthesizerListener;

/**
 * Created by damly on 2018/1/17.
 */

public class IflytekSynthesizerModule extends ReactContextBaseJavaModule {
    private static SpeechSynthesizer mTts;
    private static SynthesizerListener mTtsListener;
    private static AudioManager mAudioManager;
    private static Context mContext;
    private static VoiceSynthesizerJsEvent mVoiceSynthesizerJsEvent;

    public IflytekSynthesizerModule(ReactApplicationContext reactContext) {
        super(reactContext);

        mContext = reactContext;
        mAudioManager = (AudioManager) reactContext.getSystemService(reactContext.AUDIO_SERVICE);
        mVoiceSynthesizerJsEvent = new VoiceSynthesizerJsEvent(reactContext);
    }

    @Override
    public String getName() {
        return "IflytekSynthesizerModule";
    }

    @ReactMethod
    public void init(String appId) {
        SpeechUtility.createUtility(mContext, SpeechConstant.APPID + "=" + appId);
        mTts = SpeechSynthesizer.createSynthesizer(mContext, null);
        mTtsListener = new SynthesizerListener() {

            @Override
            public void onSpeakBegin() {
                mVoiceSynthesizerJsEvent.onStart("iflytek");
            }

            @Override
            public void onBufferProgress(int i, int i1, int i2, String s) {
            }

            @Override
            public void onSpeakPaused() {
            }

            @Override
            public void onSpeakResumed() {
            }

            @Override
            public void onSpeakProgress(int i, int i1, int i2) {
            }

            @Override
            public void onCompleted(SpeechError speechError) {
                mVoiceSynthesizerJsEvent.onCompleted();
            }

            @Override
            public void onEvent(int i, int i1, int i2, Bundle bundle) {
            }
        };
    }

    @ReactMethod
    public void start(String content, String langCode, int pronounce, String speaker) {
        if (mTts.isSpeaking()) {
            mTts.stopSpeaking();
        }

        this.setTtsParam(langCode, pronounce, speaker);

        mTts.startSpeaking(content, mTtsListener);
    }

    @ReactMethod
    public void stop() {
        if (mTts.isSpeaking()) {
            mTts.stopSpeaking();
        }
    }

    @ReactMethod
    public void isSpeaking(Promise promise) {
        promise.resolve(mTts.isSpeaking());
    }

    private void setTtsParam(String langCode, int pronounce, String speaker) {

        // 清空参数
        mTts.setParameter(SpeechConstant.PARAMS, null);

        mTts.setParameter(SpeechConstant.ENGINE_TYPE, SpeechConstant.TYPE_CLOUD);
        // 设置在线合成发音人
        mTts.setParameter(SpeechConstant.VOICE_NAME, speaker);
        //设置合成语速
        mTts.setParameter(SpeechConstant.SPEED, "50");
        //设置合成音调
        mTts.setParameter(SpeechConstant.PITCH, "50");
        //设置合成音量
        mTts.setParameter(SpeechConstant.VOLUME, "50");

        // 设置播放合成音频打断音乐播放，默认为true
        mTts.setParameter(SpeechConstant.KEY_REQUEST_FOCUS, "true");

        //设置播放器音频流类型
        if (mAudioManager.isBluetoothScoOn())
            mTts.setParameter(SpeechConstant.STREAM_TYPE, "" + AudioManager.STREAM_VOICE_CALL);
        else
            mTts.setParameter(SpeechConstant.STREAM_TYPE, "" + AudioManager.STREAM_MUSIC);
    }
}

File: ios/IflytekSynthesizerModule.m
//
//  IflytekSynthesizerModule.m
//  VoiceManager
//
//  Created by damly on 2018/1/19.
//  Copyright © 2018年 damly. All rights reserved.
//
#import "PcmPlayer.h"
#import "VoiceSynthesizerJsEvent.h"
#import "IflytekSynthesizerModule.h"
#import <Foundation/Foundation.h>


@implementation IflytekSynthesizerDelegate
{
    onFinish    _onFinish;
    onError     _onError;
    NSMutableData *_data;
}

- (id) initWithBlocks:(onFinish) onFinish onError:(onError) onError
{
    self = [super init];
    
    _onFinish = onFinish;
    _onError = onError;
    
    return self ;
}
-(id) init
{
    return [self initWithBlocks:nil onError:nil] ;
}

- (void) onCompleted:(IFlySpeechError *)error
{
    if (error != nil && error.errorCode != 0 && _onError != nil) {
        _onError(error.errorCode, error.errorDesc);
    }
    else if(_onFinish != nil){
        _onFinish();
    }
}

@end

@implementation IflytekSynthesizerModule
{
    VoiceSynthesizerJsEvent *_voiceSynthesizerJsEvent;
    IFlySpeechSynthesizer * _iFlySpeechSynthesizer;
    IflytekSynthesizerDelegate *_iflytekSynthesizerDelegate;
    PcmPlayer       *_pcmPlayer;
    NSString    *_content;
    NSString    *_filename;
    BOOL        _speaking;
}

@synthesize bridge = _bridge;

RCT_EXPORT_MODULE(IflytekSynthesizerModule);

RCT_EXPORT_METHOD(init: (NSString *) AppId) {
    
    if (_iFlySpeechSynthesizer != nil) {
        [_voiceSynthesizerJsEvent onError:-1 message:@"please init iflytek synthesizer."];
        return;
    }
    
    _voiceSynthesizerJsEvent = [[VoiceSynthesizerJsEvent alloc] initWith:_bridge];
    _pcmPlayer = [[PcmPlayer alloc] initWithBlocks:^() {
        
    } onFinish:^(){
        _speaking = NO;
        [_voiceSynthesizerJsEvent onCompleted];
    } onError:^(int code ,NSString *message) {
        [_voiceSynthesizerJsEvent onError:code message:message];
    }];
    
    _iflytekSynthesizerDelegate = [[IflytekSynthesizerDelegate alloc] initWithBlocks:^(){
        [_pcmPlayer startPlaying:_filename];
    } onError:^(int code, NSString *message) {
        [_voiceSynthesizerJsEvent onError:code message:message];
        return;
    }];
    
    NSString * initIFlytekString = [[NSString alloc] initWithFormat: @"appid=%@", AppId];
    
    [IFlySpeechUtility createUtility: initIFlytekString];
    
    _iFlySpeechSynthesizer = [IFlySpeechSynthesizer sharedInstance];
    _iFlySpeechSynthesizer.delegate = _iflytekSynthesizerDelegate;
}

RCT_EXPORT_METHOD(start:(NSString *)content langCode:(NSString *)langCode pronounce:(int)pronounce speaker:(NSString *)speaker)
{
    if ([_iFlySpeechSynthesizer isSpeaking]) {
        [_iFlySpeechSynthesizer stopSpeaking];
    }
    _content = content;

    [_voiceSynthesizerJsEvent onStart:@"Iflytek"];
    
    [self synthesizeToFile:content langCode:langCode pronounce:pronounce speaker:speaker];
    // [_iFlySpeechSynthesizer startSpeaking:content];
}

- (void) synthesizeToFile:(NSString *)content langCode:(NSString *)langCode pronounce:(int)pronounce speaker:(NSString *)speaker
{
    if ([_iFlySpeechSynthesizer isSpeaking]) {
        [_iFlySpeechSynthesizer stopSpeaking];
    }
    _content = content;
    _filename = [self getFullPath: @"tts.pcm"];
    
    // 清空参数
    [_iFlySpeechSynthesizer setParameter: nil forKey: [IFlySpeechConstant PARAMS]];
    
    [_iFlySpeechSynthesizer setParameter: [IFlySpeechConstant TYPE_CLOUD] forKey: [IFlySpeechConstant ENGINE_TYPE]];
    // 设置在线合成发音人
    [_iFlySpeechSynthesizer setParameter: speaker forKey: [IFlySpeechConstant VOICE_NAME]];
    //设置合成语速
    [_iFlySpeechSynthesizer setParameter: @"50" forKey: [IFlySpeechConstant SPEED]];
    //设置合成音调
    [_iFlySpeechSynthesizer setParameter: @"50" forKey: [IFlySpeechConstant PITCH]];
    //设置合成音量
    [_iFlySpeechSynthesizer setParameter: @"50" forKey: [IFlySpeechConstant VOLUME]];

    [_iFlySpeechSynthesizer setParameter:@"16000" forKey:[IFlySpeechConstant SAMPLE_RATE]];
    
    [_iFlySpeechSynthesizer setParameter:@"tts.pcm" forKey: [IFlySpeechConstant TTS_AUDIO_PATH]];

   // [_iFlySpeechSynthesizer startSpeaking:_content];
    [_iFlySpeechSynthesizer synthesize: _content toUri: _filename];
}

RCT_EXPORT_METHOD(stop) {

}

RCT_EXPORT_METHOD(isSpeaking: (RCTPromiseResolveBlock) resolve
                  rejecter: (RCTPromiseRejectBlock) reject) {
    BOOL isSpeaking = [_iFlySpeechSynthesizer isSpeaking];
    resolve([NSNumber numberWithBool: isSpeaking]);
}

- (NSString *)getFullPath:(NSString *)fileName
{
    NSString *full = [self getPathForDirectory:NSCachesDirectory];
    
    full = [full stringByAppendingString:@"/"];
    if(fileName == nil) {
        full = [full stringByAppendingString:@"tts.pcm"];
    }
    else
    {
        full = [full stringByAppendingString:fileName];
    }
    
    return full;
}

- (NSString *)getPathForDirectory:(int)directory
{
    NSArray *paths = NSSearchPathForDirectoriesInDomains(directory, NSUserDomainMask, YES);
    return [paths firstObject];
}

@end
